### PR TITLE
Edit survey task characteristics import for existing value change

### LINKED ISSUE #5646: Bug in Survey Task Characteristics CSV Upload
## Expected behavior

When uploading characteristics CSV to survey task, values in newly uploaded file should overwrite existing values, even if characteristic-value pair (e.g., Color=Blue) already exist.

## Current behavior

As reported by [FISHstory](https://www.zooniverse.org/projects/safmcadmin/fishstory) project team by email during beta review: if a characteristics CSV is uploaded, then uploaded again with different Y/N values, the changes are not correctly adopted -- new Y characteristics are added, but changes from Y-to-N are not made.

## Steps to replicate

I was able to replicate this behavior using survey task on a test workflow (https://www.zooniverse.org/lab/8749/workflows/11387): see T0 and use the following characteristic CSV files to upload characteristics v1 ([link to CSV](https://www.dropbox.com/s/7tivh24q1v6gwmi/birdspecies_characteristics_1.csv?dl=0)), upload characteristics v2 ([link to CSV](https://www.dropbox.com/s/f75kmfm15l0jttk/birdspecies_characteristics_2.csv?dl=0)), and clear current choices ([link to CSV](https://www.dropbox.com/s/6vak0ezb7z4lu47/birdspecies_characteristics_clear.csv?dl=0)) to reset and try again.

## Additional information

I think the solution is to remove this conditional
https://github.com/zooniverse/Panoptes-Front-End/blob/106444641e596b5e491a34ae692341b86b75c486/app/classifier/tasks/survey/editor.cjsx#L448-L449
and always call the command in the second line.

I'm assigning @mcbouslog here as he was the one to last change the relevant code as part of https://github.com/zooniverse/Panoptes-Front-End/pull/2894.

### PR DESCRIPTION
**Pull request by mcbouslog**
Staging branch URL: https://pr-5682.pfe-preview.zooniverse.org

Fixes #5646.

Describe your changes.
- for each row/choice (i.e. animal) in characteristics CSV, get characteristic value index (i.e. index of "red" in "color" characteristic array, `-1` if not in array)
- if characteristic is set (i.e. "Y") and not already included for choice's characteristics, then add it
- **if characteristic is not set (i.e. "N") but it is in choice's characteristics, then remove it**

Noting CSVs included in task instructions - http://bit.ly/1QSpevJ

__This needs some more testing to confirm not breaking existing survey task__
- while troubleshooting I've focused on animal related survey tasks, are there other types of survey tasks this can be tested against?

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by shaunanoordin** [comment]
@mcbouslog Thank you for getting a fix in for this! I'm still trying to set up new projects to thoroughly test functionality (Survey Tasks are a pain 😢) so for now I can only provide comments on the code, which unfortunately is... you know. ☕✍️ (CoffeeScript is a mega pain 😭)

Quick notes:

- good gods I just realised there's a lot of in-place array editing, like `@state.myArray[index] = 'foobar'` or `@state.myArray.splice('foo', 1)`. I mean, that _works_ in JavaScript, but in React, the `@state` won't recognise that `myArray` has changed values. Probably out of the scope of this PR, but worth noting if you're wondering "Wait why isn't the page reflecting the changes caused by my interaction until I refresh the page?"

Can you talk me through the code below a bit, please? 

```
if isSet and valueIdIndex is -1
  @state.task.choices[choiceID]?.characteristics[characteristicID].push valueID
if not isSet and valueIdIndex >= 0
  @state.task.choices[choiceID]?.characteristics[characteristicID].splice valueID, 1
```

- Specifically, can you tell me what is `isSet`? Because my brain jammed trying to parse the CoffeeScript that is `for charKeyVal, isSet of row when charKeyVal isnt 'name'` 😖 

Let me take a stab at understanding that code block...

- The first `if` case says... if the _new data for species X_ has `pattern = [ 'stripe' ]` (is that what isSet is?), but _the current data for species X_ does NOT have `pattern.contains('stripe')`, then push `"stripe"` into the `pattern` array?

- The second `if` case says... if the _new data for species X_ does NOT have `pattern = [ 'stripe' ]`, but the _current data for species X_ has `pattern.contains('stripe')`, then REMOVE `"stripe"` from the `pattern` array?

Is my understanding correct so far?

- ⚠️ Wait, if the second `if` case is attempting to remove an item for an array, is the `Array.splice()` being used correctly? I believe the format is [`Array.splice(startIndex:number, deleteCount:number)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice), but the value for `startIndex` is `valueID`, which is a string?

```
// i.e. instead of...

if not isSet and valueIdIndex >= 0
  @state.task.choices[choiceID]?.characteristics[characteristicID].splice valueID, 1

// shouldn't it be...?

if not isSet and valueIdIndex >= 0
  @state.task.choices[choiceID]?.characteristics[characteristicID].splice valueIdIndex, 1
```

Misc notes:

- For `@state.task.choices[choiceID]?.characteristics[characteristicID].push`, do you need `?.push` ?

- For `@state.task.choices[choiceID]?.characteristics[characteristicID].splice`, do you need `?.splice` ?

**Comment by mcbouslog** [comment]
@shaunanoordin thank you great comments/questions!

I think it's helpful to take a step back and see if I can describe some pieces to this to confirm my own understanding:
- when any CSV, including a characteristics CSV is imported, it iterates through each row to process the data through a few methods, including but not limited to `handleFile`, `readFile`, and `parseFileContent`
- in `parseFileContent` it might read the value of a cell with `[determineBoolean](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/survey/editor.cjsx#L379)` and as relates to the CSVs we've been using to test this bug converts a "Y" to `true` or a "N" to `false`
- `addCharacteristics` gets each row as a param, which at this point could look like:
```
{
  "name": "Duck",
  "Size=Big;": true,
  "Size=Moderate;": false,
  "Size=Small;": false
}
```
which ([ignoring name, per `isnt name`)](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/survey/editor.cjsx#L425)) gives us the `charKeyValue` ("Size=Big", "Size=Moderate") and `isSet` (`true`, `false`) in [this line of code](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/survey/editor.cjsx#L425)
- `addCharactertistics` further splits the `row` object provided, like [here](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/survey/editor.cjsx#L426-L427), notably creating `characteristicLabel` ("Size") and `valueLabel` ("Big")

previously:
- if the characteristic value is true (`isSet` = `true`) and the value ("Big") isn't already in the choice's related characteristic ("Size") array, it'd add that characteristic value ("Big") to that choice's related characteristic ("Size") array

changes:
- the above still applies, new lines 446-450
- if the characteristic value is `false` and the value **_is_** in the choice's characteristic's array (that's a problem), then it's removed from the choice's related characteristic array 

I'll attempt to answer some specific questions:

(additional line noted for variable reference)
`valueIdIndex = @state.task.choices[choiceID]?.characteristics[characteristicID].indexOf valueID`
in choice "Duck", characteristic "Size" array, what index is value "Big"? if not in array, then index is -1 
`if isSet and valueIdIndex is -1`
if the characteristic value is truthy ("Y") and the value ("Big") isn't already included in the choice's related characteristic ("Size") array then
  `@state.task.choices[choiceID]?.characteristics[characteristicID].push valueID`
  choice "Duck" characteristic "Size" array gets "Big" added to it
`if not isSet and valueIdIndex > -1`
if the characteristic value is falsey ("N") and the value ("Small") **_is included_** (not good!) in the choice's related characteristic ("Size") array then
  `choiceCharacteristicArray.splice valueIdIndex, 1`
  `@state.task.choices[choiceID]?.characteristics[characteristicID] = choiceCharacteristicArray`
  choice "Duck" characteristic "Size" array removes "Small", updates state

⚠️ Wait, if the second if case is attempting to remove an item for an array, is the Array.splice() being used correctly? I believe the format is Array.splice(startIndex:number, deleteCount:number), but the value for startIndex is valueID, which is a string?
// i.e. instead of...

if not isSet and valueIdIndex >= 0
  @state.task.choices[choiceID]?.characteristics[characteristicID].splice valueID, 1

// shouldn't it be...?

if not isSet and valueIdIndex >= 0
  @state.task.choices[choiceID]?.characteristics[characteristicID].splice valueIdIndex, 1

**YES!, changed accordingly**

File: app/classifier/tasks/survey/editor.cjsx [hunk 1]
```diff
@@ -443,10 +443,14 @@ module.exports = createReactClass
...
-        if @state.task.choices[choiceID]?.characteristics[characteristicID].indexOf(valueID) is -1
-          @state.task.choices[choiceID]?.characteristics[characteristicID].push valueID
+      choiceID = @makeID row.name
+      choiceCharacteristicArray = @state.task.choices[choiceID]?.characteristics[characteristicID]
+      valueIdIndex = choiceCharacteristicArray.indexOf valueID
+      if isSet and valueIdIndex is -1
+        @state.task.choices[choiceID]?.characteristics[characteristicID]?.push valueID
+      if not isSet and valueIdIndex > -1
+        choiceCharacteristicArray.splice valueIdIndex, 1
+        @state.task.choices[choiceID]?.characteristics[characteristicID] = choiceCharacteristicArray
```
Comment (mcbouslog) [inline app/classifier/tasks/survey/editor.cjsx:453]: Lines 452 and 453 can likely be combined, don't think `choiceCharacteristicArray` really necessary anywhere. I'll clean up this afternoon, but I think underlying logic ok.

**Comment by shaunanoordin** [comment]
@mcbouslog Thanks for the explanation! 👍 That lines up with what I'm expecting for how the code should work, so that's excellent - the code logic seems, well, pretty logical.

Re: your comments on `choiceCharacteristicArray`, you're right, Lines 452 and 453 (from commit 432b38fa1ac8450623d355e991fd2a5b64b681ea) can be combined. `choiceCharacteristicArray` isn't necessary for this PR's intent. 👌

**Comment by mcbouslog** [comment]
Thank you @shaunanoordin!!

@lcjohnso could you test this out (https://pr-5682.pfe-preview.zooniverse.org/) as well? This code is untested and rarely edited, good to get a few eyes on.

**Pull request merged**